pkg/clusteroperator/clusteroperators.go

```diff
@@ -0,0 +1,257 @@
+package clusteroperator
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"fmt"
+)
+
+// CoListBuilder provides struct for ClusterOperator List.
+type CoListBuilder struct {
+	// ClusterOperatorList definition. Used to create a clusteroperator object list
+	Definition []*ClusterOperatorBuilder
+	// Created clusteroperator object list
+	ObjectList []*ClusterOperatorBuilder
+	// apiClient opens api connection to the cluster.
+	apiClient *clients.Settings
+	// Used in functions that define or mutate ClusterOperator definition. errorMsg is processed before the
+	// ClusterOperator object is created.
+	errorMsg string
+}
+
+// GetList fetches the defined ClusterOperator list from the cluster.
+func GetList(apiClient *clients.Settings) (*CoListBuilder, error) {
+	glog.V(100).Infof("Getting ClusterOperator list")
+
+	coList, err := apiClient.ClusterOperators().List(
+		context.TODO(), metaV1.ListOptions{})
+
+	if err != nil {
+		glog.V(100).Infof("Error to list ClusterOperators")
+
+		return nil, err
+	}
+
+	var coObjects []*ClusterOperatorBuilder
+
+	for _, clusteroperator := range coList.Items {
+		copiedCo := clusteroperator
+		coBuilder := &ClusterOperatorBuilder{
+			apiClient:  apiClient,
+			Object:     &copiedCo,
+			Definition: &copiedCo,
+		}
+
+		coObjects = append(coObjects, coBuilder)
+	}
+
+	coListBuilder := &CoListBuilder{
+		apiClient:  apiClient,
+		ObjectList: coObjects,
+		Definition: coObjects,
+	}
+
+	return coListBuilder, err
+}
+
+// Discover method gets the ClusterOperators list in cluster and stores them in the builder struct.
+func (builder *CoListBuilder) Discover() error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof("Getting the list of ClusterOperator objects on this cluster")
+
+	builder.ObjectList = nil
+
+	coList, err := builder.apiClient.ClusterOperators().List(
+		context.TODO(), metaV1.ListOptions{})
+
+	if err != nil {
+		glog.V(100).Infof("Error to list ClusterOperators")
+
+		return err
+	}
+
+	if len(coList.Items) < 1 {
+		glog.V(100).Infof("Cluster doesn't have ClusterOperator deployed, bad cluster state ...")
+
+		return fmt.Errorf("ClusterOperator list is empty")
+	}
+
+	for _, clusterOperator := range coList.Items {
+		copiedCo := clusterOperator
+		coBuilder := &ClusterOperatorBuilder{
+			apiClient:  builder.apiClient,
+			Object:     &copiedCo,
+			Definition: &copiedCo,
+		}
+
+		builder.ObjectList = append(builder.ObjectList, coBuilder)
+	}
+
+	return err
+}
+
+// WaitToBeAllAvailableFor waits for all ClusterConfigs to be available for a time duration up to the timeout.
+func (builder *CoListBuilder) WaitToBeAllAvailableFor(availableDuration time.Duration,
```

Comment:
Nesting wait functionality and allowing different timeouts may get a bit confusing but if you need to do it, can you validate that  availableDuration is less than timeout? It may be easier to just accept timeout and define availableDuration internally as `availableDuration := time.Second * time.Duration(timeout.Seconds()/float64(len(builder.ObjectList)))`

Reply:
I removed the second one, it's not needed in this case